keps/sig-apps/20190412-pdb-for-scale-subresource.md

```diff
@@ -0,0 +1,84 @@
+
+### Goals
+
+- Implement support for scale subresources in PDBs
+- Avoid major performance impact from the change
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+In the current implementation, the PDB controller identifies the workload controller by going through all the pods covered by the PDB, and for each pod, check in sequence for each of the four workload controllers. Each check looks at the controller reference, and if the kind is correct, looks up the controller from the shared informer. The PDB controller then looks up the desired number of replicas from the identified controller. If the set of pods covered by the PDB identfies more than one controller, that is considered an error. 
+
+The plan is to keep this approach, but check for the scale subresource if neither of the kubernetes workload controllers match. Each of the kubernetes workload controllers actually implement the scale subresource, so it is possible to only use that one. But since shared informers can not be used with the scale subresource, we want to rely on the informers for checking the kubernetes workload controllers, and only try the scale subresource if none of the them matches. This way the controller will only need to hit the apiserver in the less likely scenario where the pods have a controller other than the standard kubernetes workload controllers.
+
+As mentioned above, the PDB controller will check the desired number of replicas from the controller for each pod covered by the PDB. In almost all scenarios the controller will be the same for all pods and therefore the desired number of replicas will be the same. This is not a concern for the kubernetes workload controllers since the controller uses shared informers. When using the scale subresource, this would lead to unnecessary calls to the apiserver. We will implement a per reconcile loop cache, so the PDB controller only needs to look up each controller once per reconcile loop.
+
+### Risks and Mitigations
+
+The major risk with this change is the additional load on the apiserver since we can't use shared informers for scale subresources. As described above, we plan to mitigate this by only checking the scale subresource when the kind of the controller ref doesn't match any of the kubernetes workload controllers, and use a cache to make sure we only need to look up each controller once per reconcile loop.
+
+## Design Details
+
+### Test Plan
+
+Additional tests are required to cover this additional feature of PDBs.
+
+### Graduation Criteria
+
+[KEP](https://github.com/kubernetes/enhancements/pull/904) for graduating PDBs to GA is already underway. It involves a change to make PDBs mutable. [PR](https://github.com/kubernetes/kubernetes/pull/69867) for this is almost ready to merge. The goal is to get both that change and this one into the next version of Kubernetes (1.15), and unless any serious issues come up, promote PDBs to GA the following release (1.16).
```

Comment:
Do you intend to gate this feature with alpha feature gate in 1.15 or release it as a beta enhancement. Since you aren't actually altering the existing behavior (you're really just adding a new code path that uses the scale subresource for non-built in types) I think the risk to existing functionality is minimal and an alpha feature gate might not be warranted.

Reply:
I don't think an alpha gate is needed for this, I have updated the KEP.